File: backend/app/domains/admin/schemas/feature_flags.py
from __future__ import annotations

from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field

# Canonical flag names. These map to settings.feature_enable_<name>.
FeatureFlagName = Literal[
    "agents",
    "mcp",
    "connectors",
    "evaluations",
    "chunking_profiles",
    "adaptive_chunking",
    "graph_rag",
    "graph_extraction",
    "graph_explorer",
    "advanced_pdf_extraction",
    "language_aware_rag",
    "pipeline_explorer",
    "local_llm_profiles",
    "experimental_profiles",
    "provider_fallback",
    "external_mcp_connectors",
    "query_rewriting",
]

ALL_FLAG_NAMES: tuple[FeatureFlagName, ...] = (
    "agents",
    "mcp",
    "connectors",
    "evaluations",
    "chunking_profiles",
    "adaptive_chunking",
    "graph_rag",
    "graph_extraction",
    "graph_explorer",
    "advanced_pdf_extraction",
    "language_aware_rag",
    "pipeline_explorer",
    "local_llm_profiles",
    "experimental_profiles",
    "provider_fallback",
    "external_mcp_connectors",
    "query_rewriting",
)


class FeatureFlagDetail(BaseModel):
    name: str
    enabled: bool
    env_default: bool
    has_org_override: bool
    override_enabled: bool | None = None
    override_reason: str | None = None
    overridden_by_user_id: str | None = None
    overridden_at: datetime | None = None


class FeatureFlagsResponse(BaseModel):
    organization_id: str
    flags: list[FeatureFlagDetail]


class FeatureFlagSetRequest(BaseModel):
    enabled: bool
    reason: str | None = Field(default=None, max_length=500)


class FeatureFlagSetResponse(BaseModel):
    organization_id: str
    flag: FeatureFlagDetail


class FeatureFlagDeleteResponse(BaseModel):
    organization_id: str
    flag_name: str
    reverted_to_env_default: bool
    env_default: bool


class PublicFeatureFlagsResponse(BaseModel):
    """Safe subset of flag states returned to authenticated non-admin users."""

    flags: dict[str, bool]

File: frontend/src/lib/api/feature-flags.ts
import { apiRequest } from "@/lib/api/request";

export type FeatureFlagName =
  | "agents"
  | "mcp"
  | "connectors"
  | "evaluations"
  | "chunking_profiles"
  | "adaptive_chunking"
  | "graph_rag"
  | "graph_extraction"
  | "graph_explorer"
  | "advanced_pdf_extraction"
  | "language_aware_rag"
  | "pipeline_explorer"
  | "local_llm_profiles"
  | "experimental_profiles"
  | "provider_fallback"
  | "external_mcp_connectors"
  | "query_rewrite_preview"
  | "org_memory";

export const ALL_FLAG_NAMES: FeatureFlagName[] = [
  "agents",
  "mcp",
  "connectors",
  "evaluations",
  "chunking_profiles",
  "adaptive_chunking",
  "graph_rag",
  "graph_extraction",
  "graph_explorer",
  "advanced_pdf_extraction",
  "language_aware_rag",
  "pipeline_explorer",
  "local_llm_profiles",
  "experimental_profiles",
  "provider_fallback",
  "external_mcp_connectors",
  "query_rewrite_preview",
  "org_memory",
];

export const FLAG_LABELS: Record<FeatureFlagName, string> = {
  agents: "Agentic mode",
  mcp: "MCP integration",
  connectors: "Connectors",
  evaluations: "Evaluations",
  chunking_profiles: "Chunking profiles",
  adaptive_chunking: "Adaptive chunking",
  graph_rag: "GraphRAG retrieval",
  graph_extraction: "Graph extraction",
  graph_explorer: "Graph explorer",
  advanced_pdf_extraction: "Advanced PDF extraction",
  language_aware_rag: "Language-aware RAG",
  pipeline_explorer: "Pipeline explorer",
  local_llm_profiles: "Local LLM profiles",
  experimental_profiles: "Experimental model profiles",
  provider_fallback: "Provider fallback",
  external_mcp_connectors: "External MCP connectors",
  query_rewrite_preview: "Query rewrite preview",
  org_memory: "Organization memory",
};

export type FeatureFlagDetail = {
  name: string;
  enabled: boolean;
  env_default: boolean;
  has_org_override: boolean;
  override_enabled: boolean | null;
  override_reason: string | null;
  overridden_by_user_id: string | null;
  overridden_at: string | null;
};

export type FeatureFlagsResponse = {
  organization_id: string;
  flags: FeatureFlagDetail[];
};

export type FeatureFlagSetRequest = {
  enabled: boolean;
  reason?: string | null;
};

export type FeatureFlagSetResponse = {
  organization_id: string;
  flag: FeatureFlagDetail;
};

export type FeatureFlagDeleteResponse = {
  organization_id: string;
  flag_name: string;
  reverted_to_env_default: boolean;
  env_default: boolean;
};

export type PublicFeatureFlagsResponse = {
  flags: Record<string, boolean>;
};

export async function listAdminFeatureFlags(): Promise<FeatureFlagsResponse> {
  return apiRequest<FeatureFlagsResponse>("/admin/feature-flags", {
    method: "GET",
  });
}

export async function setAdminFeatureFlag(
  flagName: string,
  payload: FeatureFlagSetRequest,
): Promise<FeatureFlagSetResponse> {
  return apiRequest<FeatureFlagSetResponse>(
    `/admin/feature-flags/${encodeURIComponent(flagName)}`,
    {
      method: "PUT",
      body: JSON.stringify(payload),
    },
  );
}

export async function clearAdminFeatureFlag(
  flagName: string,
): Promise<FeatureFlagDeleteResponse> {
  return apiRequest<FeatureFlagDeleteResponse>(
    `/admin/feature-flags/${encodeURIComponent(flagName)}`,
    {
      method: "DELETE",
    },
  );
}

export async function getPublicFeatureFlags(): Promise<PublicFeatureFlagsResponse> {
  return apiRequest<PublicFeatureFlagsResponse>("/feature-flags", {
    method: "GET",
  });
}
